### PR TITLE
Add: canJoin + student.Instructing + Update: pupil join/leave Subcourse

### PR DESCRIPTION
**Pull request by Leo310**

### CODE REVIEW

**Comment by Leo310** [comment]
Maybe should have pushed to a new branch

File: common/courses/participants.ts [hunk 1]
```diff
@@ -117,9 +118,39 @@ export function canJoinSubcourses(pupil: Pupil): Decision<CourseDecision> {
...
+    });
+    if (firstLecture.length !== 1) {
+        return { allowed: false, reason: 'no-lectures' };
+    }
+    if (firstLecture[0].start < new Date() && !subcourse.joinAfterStart) {
+        return { allowed: false, reason: 'already-started' };
+    }
+
+    const pupilGrade = gradeAsInt(pupil.grade);
+    if (subcourse.minGrade && subcourse.maxGrade && subcourse.minGrade < pupilGrade && pupilGrade < subcourse.maxGrade) {
```
Comment (Jonasdoubleyou) [inline common/courses/participants.ts:139]: This line is flawed. I guess

if (subcourse.minGrade && subcourse.minGrade < grade) 
  return { allowed: false };

if (subcourse.maxGrade && subcourse.maxGrade > grade)
  return { allowed: false };

is easier to reason about.

File: common/courses/participants.ts [hunk 2]
```diff
@@ -117,9 +118,39 @@ export function canJoinSubcourses(pupil: Pupil): Decision<CourseDecision> {
...
+        return { allowed: false, reason: 'inappropriate-grade' };
+    }
+
+    const participants = await prisma.subcourse_participants_pupil.count({ where: { subcourseId: subcourse.id } });
+    if (subcourse.maxParticipants < participants) {
```
Comment (Jonasdoubleyou) [inline common/courses/participants.ts:144]: <= ?

File: graphql/subcourse/mutations.ts [hunk 3]
```diff
@@ -232,4 +231,56 @@ export class MutateSubcourseResolver {
         logger.info(`Lecture (${lecture.id}) was deleted`);
         return true;
     }
+
+    @Mutation((returns) => Boolean)
+    @Authorized(Role.ADMIN, Role.PUPIL)
+    async pupilJoinSubcourse(
```
Comment (Jonasdoubleyou) [inline graphql/subcourse/mutations.ts:237]: Rename: subcourseJoin, subcourseLeave and so on

File: common/courses/participants.ts (hunk 1 above)
Comment (Leo310) [reply common/courses/participants.ts:139]: ok

File: common/courses/participants.ts (hunk 2 above)
Comment (Leo310) [reply common/courses/participants.ts:144]: true

**Review by Jonasdoubleyou — APPROVED** [review]
Don't forget to adapt the Retool (I guess just the "Support Schüler" is affected)

**Pull request merged**